File: 16-Trie/02-XOR-Trie/cpp/xor_trie.cpp
/* Maximum XOR using Trie | LeetCode 421 | Tags: xor-trie, bit-trie */
#include <bits/stdc++.h>
using namespace std;
struct TrieNode{TrieNode*ch[2]={};};
class XorTrie{TrieNode*root;public:XorTrie(){root=new TrieNode();}
void insert(int n){auto node=root;for(int i=31;i>=0;i--){int bit=(n>>i)&1;if(!node->ch[bit])node->ch[bit]=new TrieNode();node=node->ch[bit];}}
int maxXor(int n){auto node=root;int res=0;for(int i=31;i>=0;i--){int bit=(n>>i)&1;if(node->ch[1-bit]){res|=(1<<i);node=node->ch[1-bit];}else node=node->ch[bit];}return res;}};
int main(){XorTrie t;vector<int>nums={3,10,5,25,2,8};for(int x:nums)t.insert(x);
int ans=0;for(int x:nums)ans=max(ans,t.maxXor(x));cout<<"Max XOR: "<<ans<<endl;} // 28

File: 16-Trie/02-XOR-Trie/java/XorTrie.java
public class XorTrie { int[][] ch; int cnt=0;
    XorTrie(){ch=new int[3200005][2];java.util.Arrays.fill(ch[0],-1);}
    void insert(int n){int node=0;for(int i=31;i>=0;i--){int bit=(n>>i)&1;if(ch[node][bit]==-1){ch[node][bit]=++cnt;java.util.Arrays.fill(ch[cnt],-1);}node=ch[node][bit];}}
    int maxXor(int n){int node=0,res=0;for(int i=31;i>=0;i--){int bit=(n>>i)&1;if(ch[node][1-bit]!=-1){res|=(1<<i);node=ch[node][1-bit];}else node=ch[node][bit];}return res;}
    public static void main(String[] args){XorTrie t=new XorTrie();int[] nums={3,10,5,25,2,8};for(int x:nums)t.insert(x);int ans=0;for(int x:nums)ans=Math.max(ans,t.maxXor(x));System.out.println("Max XOR: "+ans);}}
